### PR TITLE
Send charge acknowledgement to post office

### PR DESCRIPTION
**Pull request by BjarkeMeier**
## Description

Contents of this pull request:

* New function `Charge Acknowledgement Sender` added
* New function subscribes to charge command accepted events and sends acknowledgements to post office
* Post office is emulated using a charges-domain-only service bus topic as there is currently no real post office and to enable the option of exploring and integration testing the charges domain without external dependencies 
* Command and event models have been refactored
* A temporary JSON deserialization solution has been made because System.Text.JsonSerializer doesn't support deserialization of types without parameterless constructor and/or without public setters. Both are needed.
* HTTP GET health endpoint has been added to the new function to allow health monitoring to be added later

Here is an example of an acknowledgement that ended up in the post office:

```json
{
  "CorrelationId": "a8a20348-68da-41c2-a8b7-262c8c98d2f0",
  "ReceiverMRid": "8100000000030",
  "ReceiverBusinessProcessRole": 0,
  "OriginalTransactionReferenceMRid": "MAR2021-05-05T14:22:37.393Z",
  "BusinessReasonCode": 18
}
```

## References

* #134

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeAcknowledgementSender/GreenEnergyHub.Charges.ChargeAcknowledgementSender.csproj [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.11" />
+    </ItemGroup>
+    <ItemGroup>
+        <None Update="host.json">
+            <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        </None>
+        <None Update="local.settings.json">
+            <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+            <CopyToPublishDirectory>Never</CopyToPublishDirectory>
+        </None>
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeAcknowledgementSender/GreenEnergyHub.Charges.ChargeAcknowledgementSender.csproj:35]: Do we want to include the "personal" local.settings.json? - in the two other functions in the Charges domain we only include the local.settings.sample.json file.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChargeAcknowledgement.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// limitations under the License.
+
+using GreenEnergyHub.Charges.Domain.Common;
+using JetBrains.Annotations;
+
+namespace GreenEnergyHub.Charges.Domain
+{
+    public class ChargeAcknowledgement
+    {
+        public ChargeAcknowledgement(string correlationId, string receiverMRid, MarketParticipantRole receiverBusinessProcessRole, object originalTransactionReferenceMRid, ProcessType businessReasonCode)
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChargeAcknowledgement.cs:22]: consider newline'ing each parameter for better viewing the file on github :)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/PostOffice/PostOfficeService.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+            var queueOrTopicName = _options.Value.TopicName;
+            ServiceBusSender sender = client.CreateSender(queueOrTopicName);
+            var serializedMessage = _jsonSerializer.Serialize(acknowledgement);
+            var message = new ServiceBusMessage(serializedMessage)
+            {
+                CorrelationId = acknowledgement!.CorrelationId,
+
+                // We set 'Subject' at the moment for a better overview in the AZ portal.
+                Subject = acknowledgement.GetType().Name,
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/PostOffice/PostOfficeService.cs:48]: We have removed our "catch all" subscriber if i remember correctly, so right now we have a 1-1 relation between events and topics. So we no longer have a place where we benefit from setting Subject as I see it.

**Review by lasrinnil — APPROVED** [review]
Great job! One step closer to a full flow within our own domain 🥳

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeAcknowledgementSender/GreenEnergyHub.Charges.ChargeAcknowledgementSender.csproj [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeAcknowledgementSender/GreenEnergyHub.Charges.ChargeAcknowledgementSender.csproj:18]: I'm not sure what you're referring to?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/ChargeCommandNullCheckerTest.cs [hunk 5]
```diff
@@ -122,12 +121,11 @@ public void ChargeCommandMktActivityRecordIsNullThrowsException()
             Assert.Throws<ArgumentNullException>(() => ChargeCommandNullChecker.ThrowExceptionIfRequiredPropertyIsNull(c));
         }
 
-        private static ChargeCommand GetValidCharge()
+        private static ChargeCommand GetValidCharge(string correlationId = "some-correlation-id")
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/ChargeCommandNullCheckerTest.cs:124]: Intentional default value or a leftover from testing new code?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeAcknowledgementSender/GreenEnergyHub.Charges.ChargeAcknowledgementSender.csproj (hunk 1 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeAcknowledgementSender/GreenEnergyHub.Charges.ChargeAcknowledgementSender.csproj:35]: I don't think I added it to source control. And I also think the other hosts refer to the file in the project file. Or am I missing something?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Core/Json/JsonSerializerTests.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+using JsonSerializer = GreenEnergyHub.Charges.Core.Json.JsonSerializer;
+
+namespace GreenEnergyHub.Charges.Tests.Core.Json
+{
+    [Trait(TraitNames.Category, TraitValues.UnitTest)]
+    public class JsonSerializerTests
+    {
+        private const string ChargeCommandAcceptedEventMRid = "MAR2021-05-05T13:56:29.301Z";
+
+        private const string ChargeCommandAcceptedEventJsonString = @"
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Core/Json/JsonSerializerTests.cs:29]: Personally I prefer to put json test data in a separate file.

**Review by HenrikSommer — APPROVED** [review]

**Comment by BjarkeMeier** [comment]
Thank you @lasrinnil and @HenrikSommer for your feedback. I should now have either fixed or replied to all of your comments. See fixes in the last commit.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeAcknowledgementSender/GreenEnergyHub.Charges.ChargeAcknowledgementSender.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeAcknowledgementSender/GreenEnergyHub.Charges.ChargeAcknowledgementSender.csproj:18]: Thats weird what it pasted was the tags for:
image

but guess the tag is removed when posting here

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/ChargeCommandNullCheckerTest.cs (hunk 5 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/ChargeCommandNullCheckerTest.cs:124]: Intentional. The method is used multiple times without arg.

**Pull request merged**